libs/grok/src/main/java/org/elasticsearch/grok/GrokCaptureConfig.java

```diff
@@ -50,18 +54,104 @@ public String name() {
     /**
      * The type defined for the field in the pattern.
      */
-    public GrokCaptureType type() {
+    GrokCaptureType type() {
         return type;
     }
 
-    Object extract(byte[] utf8Bytes, int offset, Region region) {
-        for (int number : backRefs) {
-            if (region.beg[number] >= 0) {
-                String matchValue = new String(utf8Bytes, offset + region.beg[number], region.end[number] - region.beg[number],
-                    StandardCharsets.UTF_8);
-                return type.parse(matchValue);
+    /**
+     * Build a {@linkplain GrokCaptureExtracter} that will call {@code emit} when
+     * it extracts text, boxed if the "native" representation is primitive type.
+     * Extracters returned from this method are stateless and can be reused.
+     */
+    GrokCaptureExtracter objectExtracter(Consumer<Object> emit) {
+        return nativeExtracter(new NativeExtracterMap<GrokCaptureExtracter>() {
+            @Override
+            public GrokCaptureExtracter forString(Function<Consumer<String>, GrokCaptureExtracter> buildExtracter) {
+                return buildExtracter.apply(str -> emit.accept(str));
+            }
+
+            @Override
+            public GrokCaptureExtracter forInt(Function<IntConsumer, GrokCaptureExtracter> buildExtracter) {
+                return buildExtracter.apply(i -> emit.accept(Integer.valueOf(i)));
             }
-        }
-        return null;
+
+            @Override
+            public GrokCaptureExtracter forLong(Function<LongConsumer, GrokCaptureExtracter> buildExtracter) {
+                return buildExtracter.apply(l -> emit.accept(Long.valueOf(l)));
+            }
+
+            @Override
+            public GrokCaptureExtracter forFloat(Function<FloatConsumer, GrokCaptureExtracter> buildExtracter) {
+                return buildExtracter.apply(f -> emit.accept(Float.valueOf(f)));
+            }
+
+            @Override
+            public GrokCaptureExtracter forDouble(Function<DoubleConsumer, GrokCaptureExtracter> buildExtracter) {
+                return buildExtracter.apply(d -> emit.accept(Double.valueOf(d)));
+            }
+
+            @Override
+            public GrokCaptureExtracter forBoolean(Function<Consumer<Boolean>, GrokCaptureExtracter> buildExtracter) {
```

Comment:
Should we also have a `BooleanConsumer`? That would prevent boolean boxing as `Boolean.parseBoolean` returns a native `boolean`. This boxing  is admittedly less of a concern than numeric unboxing/reboxing, but it would be nice to have consistency.

Reply:
Java didn't add one of those because they figured the boxing overhead was fairly small and they'd never allocate when boxing a boolean. I *think* that's fine for us here too. We might see otherwise, but I figure it'd be ok for now.

Reply:
👍